pool.go

```diff
@@ -37,31 +37,97 @@ type pool interface {
 	Stats() *PoolStats
 }
 
-type connList struct {
+// connStack is used as a LIFO to maintain free connection
+type connStack struct {
 	cns  []*conn
+	free chan struct{}
 	mx   sync.Mutex
-	len  int32 // atomic
-	size int32
 }
 
-func newConnList(size int) *connList {
+func newConnStack(max int) *connStack {
+	return &connStack{
+		cns:  make([]*conn, 0, max),
+		free: make(chan struct{}, max),
+	}
+}
+
+func (s *connStack) Len() int { return len(s.free) }
+
+func (s *connStack) Push(cn *conn) {
+	s.mx.Lock()
+	s.cns = append(s.cns, cn)
+	s.mx.Unlock()
+	s.free <- struct{}{}
+}
+
+func (s *connStack) ShiftStale(timeout time.Duration) *conn {
+	select {
+	case <-s.free:
+		s.mx.Lock()
+		defer s.mx.Unlock()
+
+		if cn := s.cns[0]; cn.IsStale(timeout) {
+			s.cns = s.cns[1:]
```

Comment:
If I understand correctly there is a memory leak here, because `s.cns` slice is slowly growing.